### PR TITLE
2810 Added ES mapping documents for Person and child objects.

### PR DESCRIPTION
**Pull request by albertisfu**
In this PR, I have added the mapping documents for `Person` and their child documents using the ES Join field approach.

This approach for nested documents requires that parent and child documents are stored in the same index. Therefore, the documents for parent and child objects are based on `PersonBaseDocument`.

After reviewing the previous SOLR approach, I noticed that `political_affiliation`s and `aba_ratings` objects are being flattened into a single field. Since join relations and queries should only be used where necessary, I believe it's better to keep them denormalized unless we want to introduce other fields for these objects to be searchable independently.

For `Positions`, multiple fields for each position are indexed, so in this case, it's worth indexing as a child object. This allows these field combinations to be searchable independently.

Currently, for `Education` objects, only the school name is indexed. However, there are other interesting fields like `degree_level`, `degree_year`, and `degree_detail` that can be displayed as being searchable. Therefore, I've added `Education` objects as child objects. If you believe these extra fields do not add value to users, let me know and I can revise the approach to index only school names as a flat field.

As such, the index structure for `Person` is as follows:

```
Person
 - Positions
 - Education
```

To index/update these parent and child documents, I needed to use custom signals to assign the correct parent objects.

I do have one question. I noticed the property `is_judge`, but I didn't see it being used to determine when to index a judge in SOLR (I only saw it in tests).

So, my question is, should we index only `Person` objects which are "Judges"? If so, I think the logic to index a `Person` should change. Instead of indexing it when the object is created, it should be indexed when a` judicial` Position is created for the `Person` object.

Also, should we index only `judicial_position` positions? Or all of them?

### CODE REVIEW

**Comment by mlissner** [comment]
Sorry to be so very slow replying to this one:

> So, my question is, should we index only Person objects which are "Judges"? If so, I think the logic to index a Person should change. Instead of indexing it when the object is created, it should be indexed when a judicial Position is created for the Person object.

That'd make sense to me, yes, but I won't block this PR on that. We rarely get new people in the system, so let's make that a tweak for later (if you didn't already include it).

> Also, should we index only judicial_position positions? Or all of them?

All of them. People should be able to search by any position a judge has.

File: cl/search/documents.py
Comment (mlissner) [inline cl/search/documents.py:460]: Can you take a look at person_text.txt and see which of those fields we actually still need? Maybe we can do without the `text` field again?

File: cl/search/es_indices.py
Comment (mlissner) [inline cl/search/es_indices.py:33]: We want distinct settings here, don't we? Like we did for OA?

File: cl/search/signals.py
Comment (mlissner) [inline cl/search/signals.py:1]: This is a little too easy, jeez.

**Review by mlissner — CHANGES_REQUESTED** [review]
A few small comments from me. I didn't see a waffle and this feels too small. I'm missing something, aren't I?

File: cl/search/documents.py
Comment (albertisfu) [reply cl/search/documents.py:460]: Yeah, `text` is already removed in #2870

File: cl/search/es_indices.py
Comment (albertisfu) [reply cl/search/es_indices.py]: Sure, I've added custom settings for People.

**Comment by albertisfu** [comment]
> That'd make sense to me, yes, but I won't block this PR on that. We rarely get new people in the system, so let's make that a tweak for later (if you didn't already include it).

Yeah, this logic is included in #2870 which should be merged here before we move this to prod.

> All of them. People should be able to search by any position a judge has.

Yeah, this is how it works in #2870 we index in elastic search only Persons that have at least one judiciary position. But once the person is indexed we continue indexing all their positions including not judiciary positions. 
A tweak that might be required is that for instance a `Person` that at the beginning only has no judiciary positions won't be indexed until a judiciary position for that person is created. But in that case, all the previous judiciary positions will be outside the index so a tweak I can apply in #2870 is to check if there are judiciary positions that belong to the Person so they can be also indexed when indexing the judiciary position.

**Comment by mlissner** [comment]
> A tweak that might be required is that for instance a Person that at the beginning only has no judiciary positions won't be indexed until a judiciary position for that person is created.

That'd be good, I think. I could see us loading a bunch of people into the DB, and then in a separate process adding their positions (judicial and otherwise). 

----

It sounds like once #2870 is merged here, this PR is ready to merge into `main`. Correct?

**Comment by albertisfu** [comment]
> That'd be good, I think. I could see us loading a bunch of people into the DB, and then in a separate process adding their positions (judicial and otherwise).

Yeah, I'm working on tweaking the signal processor in #2870 so this can be possible.


It sounds like once https://github.com/freelawproject/courtlistener/pull/2870 is merged here, this PR is ready to merge into main. Correct?

Yeah, that's correct, I'll let you know when #2870 is ready.

**Comment by albertisfu** [comment]
> For the indexing question, we could do a hack. There aren't that many people that aren't judges. Could we just add everything and then do some quick custom code to remove folks we don't want in the index?

Yes, I've checked this, and it would be possible to remove docs that are not judges from the index afterward. The main issue is that child objects require a `_routing` parameter during indexing, which is the ID of their parent document. Since built-in indexing commands use bulk indexing, I don't see an easy way to incorporate this parameter into document mapping for being automatically assigned during indexing.

So we would still need a custom command for indexing Positions. Given that there aren't many judges, I believe we could do it easily by using Celery tasks to index both the Judges and their Positions, similar to the command we wrote for indexing alerts.

For RECAP and Opinions, this command may require performance improvements, which we could assess after evaluating how the indexing of Judges goes.

**Comment by mlissner** [comment]
Sounds good. I guess we'll get experience we'll need for indexing RECAP and opinions.

**Comment by albertisfu** [comment]
Okay, this is ready to be deployed! I've just pushed a change that unifies the flags. Previously, we had two different flags: `p-es-activate` for the search API and `p-es-active` for the front end. It's better to have just one, `p-es-active`.

So the process we should follow to get this landed is the following.

Before merging to main:

Set the number of shards and replicas for the Person Search index env vars:

```
ELASTICSEARCH_PEOPLE_NUMBER_OF_SHARDS
ELASTICSEARCH_PEOPLE_NUMBER_OF_REPLICAS
```

Then we can merge and deploy and

Create the index:
`search_index --create --models people_db.Person`

If it throws an error `resource_already_exists_exception`, it means that a Judge has just been created and the index has been automatically generated. However, this automatically generated index is not compatible with parent-child objects. Therefore, it needs to be removed and recreated.

`search_index --delete --models people_db.Person`
`search_index --create --models people_db.Person`


And finally, we can index Judges and their positions:
`manage.py cl_index_judges_and_positions --pk-offset 0 --queue celery`

**Review by mlissner — APPROVED** [review]

**Pull request merged**
